flow-server/src/main/java/com/vaadin/flow/server/VaadinSession.java

```
@@ -1018,6 +1026,11 @@ public String getPushId() {
      */
     private void readObject(ObjectInputStream stream)
             throws IOException, ClassNotFoundException {
+        boolean wasSerialized = stream.readBoolean();
+        this.deserializedAsEmpty = !wasSerialized;
+        if (!wasSerialized) {
```

I don't understand this code.
It just returns if serialization has not happened.

It means that the state of the object is not restored proeprly.
It means e.g. that transient `pendingAccessQueue` has `null` value which means that the code becomes broken.

So: shouldn't this code just thrown an exception if an object is deserialized being not serialized ?

---

The whole change is about conditionally not including `VaadinSession` in the serialized HTTP session.

If you throw an exception during serialization or deserialization, then you abort the whole process.

When the session is deserialized, it is marked as "empty" and then ignored and removed by `VaadinService` when encountered.

---

Would be good to have this in the comments.

Another thing which I would like to see here: protection from using the instance which is not properly deserialized.
It should make itself as invalid (via some flag or somehow else) and then every access to the instance should throw an exception which says that the instance is invalid: a new one should have been created and used instead.

---

Added an assert to `checkHasLock` which is called from most methods

---

Added javadoc